### PR TITLE
Custom schema validation

### PR DESCRIPTION
**Pull request by KavishShah09**

### CODE REVIEW

File: gateway/modules/eventing/helpers.go [hunk 1]
```diff
@@ -228,4 +242,16 @@ func (m *Module) selectRule(name string) (config.EventingRule, error) {
...
+
+func (m *Module) validate(event *model.QueueEventRequest) error {
+	if event.Type == utils.EventDBCreate || event.Type == utils.EventDBDelete || event.Type == utils.EventDBUpdate || event.Type == utils.EventFileCreate || event.Type == utils.EventFileDelete {
+		return nil
+	}
+	schema, p := m.schemas[event.Type]
+	if !p && m.config.IsForce == true {
+		return errors.New("required key not present in request")
+	}
+	_, err := m.schema.SchemaValidator(event.Type, schema, event.Payload.(map[string]interface{}))
```
Comment (YourTechBud) [inline gateway/modules/eventing/helpers.go]: The logic is incorrect
- Return error if IsForce is true and key isn't present
- Return nil if key isn't present
- Validate schema if key is present

File: gateway/modules/schema/schema.go [hunk 2]
```diff
@@ -94,8 +95,10 @@ func (s *Schema) parser(crud config.Crud) (schemaType, error) {
 			// parse the source
 			doc, err := parser.Parse(parser.ParseParams{Source: source})
 			if err != nil {
+				log.Println("error", err)
```
Comment (YourTechBud) [inline gateway/modules/schema/schema.go]: Remove this

File: gateway/modules/schema/schema.go [hunk 3]
```diff
@@ -94,8 +95,10 @@ func (s *Schema) parser(crud config.Crud) (schemaType, error) {
...
 				return nil, err
 			}
+			log.Println("parsed below")
```
Comment (YourTechBud) [inline gateway/modules/schema/schema.go]: remove this

File: gateway/utils/handlers/config_eventing.go [hunk 4]
```diff
@@ -119,3 +119,81 @@ func HandleSetEventingConfig(adminMan *admin.Manager, syncMan *syncman.Manager)
 
 	}
 }
+
```
Comment (YourTechBud) [inline gateway/utils/handlers/config_eventing.go]: Create another endpoint to set the `IsForce` / `strict` value

File: gateway/utils/handlers/config_eventing.go [hunk 5]
```diff
@@ -119,3 +119,81 @@ func HandleSetEventingConfig(adminMan *admin.Manager, syncMan *syncman.Manager)
...
+		if err := syncMan.SetEventingSchema(ctx, project, c.Type, c.Schema); err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		w.WriteHeader(http.StatusOK) //http status codee
+		json.NewEncoder(w).Encode(map[string]interface{}{})
+		return
+
```
Comment (YourTechBud) [inline gateway/utils/handlers/config_eventing.go]: Remove line

File: gateway/utils/server/routes.go [hunk 6]
```diff
@@ -31,6 +31,8 @@ func (s *Server) routes(profiler bool, staticPath string) *mux.Router {
 	router.Methods("POST").Path("/v1/config/projects/{project}/event-triggers/rules/{ruleName}").HandlerFunc(handlers.HandleAddEventingRule(s.adminMan, s.syncMan))
 	router.Methods("DELETE").Path("/v1/config/projects/{project}/event-triggers/rules/{ruleName}").HandlerFunc(handlers.HandleDeleteEventingRule(s.adminMan, s.syncMan))
 	router.Methods("POST").Path("/v1/config/projects/{project}/event-triggers/config").HandlerFunc(handlers.HandleSetEventingConfig(s.adminMan, s.syncMan))
+	router.Methods("POST").Path("/v1/config/projects/{project}/event-triggers/schema").HandlerFunc(handlers.HandleSetEventingSchema(s.adminMan, s.syncMan))
+	router.Methods("DELETE").Path("/v1/config/projects/{project}/event-triggers/schema").HandlerFunc(handlers.HandleDeleteEventingSchema(s.adminMan, s.syncMan))
```
Comment (YourTechBud) [inline gateway/utils/server/routes.go]: Add one more route to set the `strict` field

File: gateway/utils/syncman/syncman_eventing.go [hunk 7]
```diff
@@ -48,3 +49,38 @@ func (s *Manager) SetEventingConfig(ctx context.Context, project, dbType, col st
...
+		return err
+	}
+	if len(projectConfig.Modules.Eventing.Schemas) != 0 {
+		projectConfig.Modules.Eventing.Schemas[evType] = config.SchemaObject{Schema: schema}
+	} else {
+		projectConfig.Modules.Eventing.Schemas = map[string]config.SchemaObject{
+			evType: config.SchemaObject{Schema: schema},
+		}
+	}
+	log.Println("checking schema in handler:", projectConfig.Modules.Eventing.Schemas)
```
Comment (YourTechBud) [inline gateway/utils/syncman/syncman_eventing.go]: Remove this log

File: gateway/utils/handlers/config_eventing.go [hunk 8]
```diff
@@ -119,3 +119,117 @@ func HandleSetEventingConfig(adminMan *admin.Manager, syncMan *syncman.Manager)
...
+	}
+}
+
+func HandleDeleteEventingSchema(adminMan *admin.Manager, syncMan *syncman.Manager) http.HandlerFunc {
+	type schemaDeleteRequest struct {
+		Type string `json:type`
```
Comment (sharadregoti) [inline gateway/utils/handlers/config_eventing.go]: add double quotes in json annotation

File: gateway/modules/eventing/handle_staged.go [hunk 9]
```diff
@@ -108,56 +109,66 @@ func (m *Module) processStagedEvent(eventDoc *model.EventDocument) {
 		Time: time.Unix(0, eventDoc.Timestamp*int64(time.Millisecond)).Format(time.RFC3339), Data: eventDoc.Payload}
 
 	for {
-		internalToken, err := m.auth.GetInternalAccessToken()
-		if err != nil {
-			log.Println("Eventing: Couldn't trigger functions -", err)
-			return
-		}
+		if err := m.invokeWebhook(ctx, rule.Timeout, eventDoc, &cloudEvent); err != nil {
+			log.Println("Eventing staged event handler could not get response from service:", err)
```
Comment (sharadregoti) [inline gateway/modules/eventing/handle_staged.go]: change all the log message with logrus

File: gateway/config/config.go [hunk 10]
```diff
@@ -177,6 +179,10 @@ type EventingRule struct {
 	Options map[string]string `json:"options" yaml:"options"`
 }
 
+type SchemaObject struct {
+	Schema string `json:"schema" yaml:"schema"`
+}
+
```
Comment (sharadregoti) [inline gateway/config/config.go]: This should be in the handler

File: gateway/modules/eventing/handle_staged.go [hunk 11]
```diff
@@ -90,15 +86,20 @@ func (m *Module) processStagedEvent(eventDoc *model.EventDocument) {
...
 	}
-	
+
+	if rule.Timeout == 0 {
+		rule.Timeout = 5000
+	}
+
 	// Call the function to process the event
-	ctxLocal, cancel := context.WithTimeout(ctx, time.Duration(maxRetries*rule.Timeout)*time.Millisecond)
+	timeoutLocal := time.Duration(maxRetries*rule.Timeout+15000) * time.Millisecond
```
Comment (sharadregoti) [inline gateway/modules/eventing/handle_staged.go]: cannot use `15000`/ it should be `5000*maxRetries`

File: gateway/modules/eventing/helpers.go [hunk 12]
```diff
@@ -228,4 +235,19 @@ func (m *Module) selectRule(name string) (config.EventingRule, error) {
...
+}
+
+func (m *Module) validate(event *model.QueueEventRequest) error {
+	if event.Type == utils.EventDBCreate || event.Type == utils.EventDBDelete || event.Type == utils.EventDBUpdate || event.Type == utils.EventFileCreate || event.Type == utils.EventFileDelete {
+		return nil
+	}
+	schema, p := m.schemas[event.Type]
+	if !p {
+		if m.config.Strict == true {
+			return errors.New("required key not present in request")
```
Comment (sharadregoti) [inline gateway/modules/eventing/helpers.go]:
```suggestion
			return fmt.Errorf("schema not found for event (%s)", event.Type)
```

File: gateway/config/config.go (hunk 10 above)
Comment (KavishShah09) [reply gateway/config/config.go]: Can not shift this since being used in Eventing struct in the same file

**Pull request merged**